labs/computing-experiment-subjects/index.md

```diff
@@ -0,0 +1,213 @@
+- In order to be counted in the analysis, a visitor must have at least one "decision" in the analysis window.
+- The variation and the timestamp assigned to that visitor correspond to the variation in the _first_ decision received from that visitor during the analysis window
+
+**Note:** the visitor counting logic used in [Optimizely Web](https://www.optimizely.com/platform/experimentation/) experiments is slightly different.  See [this article](https://help.optimizely.com/Analyze_Results/How_Optimizely_counts_conversions) for details.
+
+
+```python
+spark.sql(f"""
+    CREATE OR REPLACE TEMPORARY VIEW experiment_subjects AS
+        SELECT
+            {decision_subject_id},
+            experiment_id,
+            variation_id,
+            timestamp
+        FROM (
+            SELECT
+                *,
+                RANK() OVER (PARTITION BY experiment_id, {decision_subject_id} ORDER BY timestamp ASC) AS rnk
+            FROM
+                decisions
+        )
+        WHERE
+            rnk = 1
+        ORDER BY timestamp ASC
+""")
+
+spark.sql("""SELECT * FROM experiment_subjects LIMIT 5""")
+```
+
+
+
+
+<table border='1'>
+<tr><th>visitor_id</th><th>experiment_id</th><th>variation_id</th><th>timestamp</th></tr>
+<tr><td>visitor_1590445653085</td><td>18156943409</td><td>18174970251</td><td>2020-05-25 15:27:33.085</td></tr>
+<tr><td>visitor_1590445653325</td><td>18156943409</td><td>18112613000</td><td>2020-05-25 15:27:33.325</td></tr>
+<tr><td>visitor_1590445653565</td><td>18156943409</td><td>18112613000</td><td>2020-05-25 15:27:33.565</td></tr>
+<tr><td>visitor_1590445653805</td><td>18156943409</td><td>18174970251</td><td>2020-05-25 15:27:33.805</td></tr>
+<tr><td>visitor_1590445654045</td><td>18156943409</td><td>18174970251</td><td>2020-05-25 15:27:34.045</td></tr>
+</table>
+
+
+
+
+## Writing our subjects dataset to disk
+
+We'll store our experiment subjects in the directory specified by `subjects_output_dir`.  Subject data is partitioned into directories for each experiment included in the input decision data.
+
+
+```python
+spark.sql("""SELECT * FROM experiment_subjects""") \
+    .coalesce(1) \
+    .write.mode('overwrite') \
+    .partitionBy("experiment_id") \
+    .parquet(subjects_output_dir)
+```
+
+## Analyzing our subject data
+
+We can group by `experiment_id` and `variation_id` to count the number of subjects attributed to each variation:
+
+
+```python
+spark.sql("""
+    SELECT 
+        experiment_id, 
+        variation_id, 
+        COUNT(1) as subject_count 
+    FROM 
+        experiment_subjects 
+    GROUP BY 
+        experiment_id,
+        variation_id
+    ORDER BY
+        experiment_id,
+        variation_id
+""")
+```
+
+
+
+
+<table border='1'>
+<tr><th>experiment_id</th><th>variation_id</th><th>subject_count</th></tr>
+<tr><td>18156943409</td><td>18112613000</td><td>4487</td></tr>
+<tr><td>18156943409</td><td>18174970251</td><td>4514</td></tr>
```

Comment:
Is this it? Maybe provide a conclusion around "We've counted the number of experiment subjects across different experiment / variations

Reply:
Added a conclusion